apps/backend/src/modules/courses/course.service.ts

```diff
@@ -1,18 +1,208 @@
+		}
+
+		return vendorModule;
+	}
+
+	private async getVendorCourse(
+		vendorCourseId: string,
+		vendorId: number,
+	): Promise<CourseEntity> {
+		const vendorApi = await this.getVendorApiById(vendorId);
+		const course = await vendorApi.getCourseById(vendorCourseId);
+
+		return CourseEntity.initializeNew({
+			description: course.description,
+			image: course.image,
+			title: course.title,
+			url: course.url,
+			vendorCourseId: course.vendorCourseId,
+			vendorId,
+		});
+	}
+
+	public async addCourse({
+		userId,
+		vendorCourseId,
+		vendorId,
+	}: {
+		userId: number;
+		vendorCourseId: string;
+		vendorId: number;
+	}): Promise<CourseDto> {
+		const vendorCourse = await this.getVendorCourse(vendorCourseId, vendorId);
+
+		const addedCourse = await this.courseRepository.addCourseToUser(
+			vendorCourse,
+			userId,
+		);
+
+		return addedCourse.toObject();
+	}
+
+	public async create({
+		vendorCourseId,
+		vendorId,
+	}: {
+		vendorCourseId: string;
+		vendorId: number;
+	}): Promise<CourseDto> {
+		const existingCourse =
+			await this.courseRepository.findByVendorCourseId(vendorCourseId);
 
-	public constructor({ udemy }: Constructor) {
-		this.udemy = udemy;
+		if (existingCourse) {
+			throw new CourseError(
+				CourseErrorMessage.COURSE_IS_ALREADY_EXISTS,
+				HTTPCode.BAD_REQUEST,
+			);
+		}
+
+		const vendorCourse = await this.getVendorCourse(vendorCourseId, vendorId);
+		const addedCourse = await this.courseRepository.create(vendorCourse);
+
+		return addedCourse.toObject();
+	}
+
+	public async delete(id: number): Promise<boolean> {
+		return await this.courseRepository.delete(id);
+	}
+
+	public async find(id: number): Promise<CourseDto> {
+		const entity = await this.courseRepository.find(id);
+
+		if (!entity) {
+			throw new CourseError(
+				CourseErrorMessage.NOT_FOUND_COURSE,
+				HTTPCode.BAD_REQUEST,
+			);
+		}
+
+		return entity.toObject();
+	}
+
+	public async findAll(): Promise<CourseDto[]> {
+		const entities = await this.courseRepository.findAll();
+
+		return entities.map((entity) => entity.toObject());
 	}
 
-	public findAllByVendor(): Promise<unknown> {
-		return this.udemy.getCourses().then((result) => result.json());
+	public async findAllByVendor(
+		search: string,
+		vendor: VendorResponseDto,
+	): Promise<CourseDto[]> {
+		const vendorApi = this.getVendorApiByKey(vendor.key);
+		const items = await vendorApi.getCourses(search);
+
+		return items.map((item) => ({ ...item, id: null, vendor }));
+	}
+
+	public async findAllByVendors(parameters: {
+		search: string;
+		userId: number;
+		vendorsKeys: string | undefined;
+	}): Promise<CoursesResponseDto> {
+		const { search, userId, vendorsKeys } = parameters;
+		const vendors = vendorsKeys
+			? await this.vendorService.findAllByKeys(vendorsKeys.split(","))
+			: await this.vendorService.findAll();
+
+		const vendorsCourses = await Promise.all(
+			vendors.map(async (vendor) => {
+				return await this.findAllByVendor(search, vendor);
+			}),
+		);
```

Comment:
```suggestion
		const vendorsCourses = await Promise.all(
			vendors.map((vendor) => {
				return this.findAllByVendor(search, vendor);
			}),
		);
```
`await Promise.all` should be enough